### PR TITLE
Feature interceptor

### PR DESCRIPTION
**Pull request by maxkimambo**
+ Adding Http interceptors to add authorization headers 
+ Proxy configuration to fix CORS errors while developing

### CODE REVIEW

File: server/config/passport-setup.ts [hunk 1]
```diff
@@ -11,69 +11,137 @@
...
+ * @return {any}
+ */
+export const isAuthenticated = (req: Request, res: Response, next: NextFunction): any => {
+    const authorizationHeader = req.headers.authorization || '';
+    const accessToken = _extractAccessToken(authorizationHeader);
+    // Allowing clients to authenticate via authorization headers.
+    // at the end check if the cookies are valid and let the request through.
+    log.debug(accessToken);
+
+    if (process.env.NODe_ENV == 'production') {
```
Comment (paranerd) [inline server/config/passport-setup.ts:68]: Typo? `NODe_ENV` -> `NODE_ENV`

File: client/angular.json [hunk 2]
```diff
@@ -71,6 +71,10 @@
         },
         "serve": {
           "builder": "@angular-devkit/build-angular:dev-server",
+          "options": {
+            "browserTarget": "client:build",
+            "proxyConfig": "src/proxy.conf.json"
```
Comment (paranerd) [inline client/angular.json:76]: Why introduce the proxy?

File: client/src/app/interceptors/auth.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  private  MAXRETRIES = 2;
+  private retries = 0;
+
+
+  constructor(private http: HttpClient) { }
+
+  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
+
+    let accessToken = '';
+    const localUser = localStorage.getItem('user');
```
Comment (paranerd) [inline client/src/app/interceptors/auth.interceptor.ts:33]: Let's use the Auth service here (inject `AuthService` and use `auth.user`, which is a getter function)

File: client/src/app/interceptors/auth.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    if (localUser) {
+      const user = User.fromJson(localUser);
```
Comment (paranerd) [inline client/src/app/interceptors/auth.interceptor.ts:35]: Using `auth.user` will make the call to `fromJson` unnecessary. Also: We already have a function `User().deserialize()`, please pull the latest code

File: client/src/app/interceptors/auth.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      accessToken = user.token.access;
+    }
+    
+    const req = request.clone({ setHeaders: { 'Authorization': 'Bearer ' + accessToken }, });
+    
+    return next.handle(req).pipe(catchError((err) => {
+      const codes = new Set([401, 403, 504]);
+      if (codes.has(err.status) && this.retries < this.MAXRETRIES) {
+
+        this.retries ++;
```
Comment (paranerd) [inline client/src/app/interceptors/auth.interceptor.ts:45]: Remove the space.

File: client/src/app/interceptors/auth.interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    return next.handle(req).pipe(catchError((err) => {
+      const codes = new Set([401, 403, 504]);
+      if (codes.has(err.status) && this.retries < this.MAXRETRIES) {
+
+        this.retries ++; 
+
+        this.refreshAccessToken().subscribe({
+          next(t) {
+            console.log(t);
+            const retryRequest = req.clone({ setHeaders: { 'Authorization': 'Bearer ' + t.access }, });
```
Comment (paranerd) [inline client/src/app/interceptors/auth.interceptor.ts:50]: I think you don't do the trailing comma when staying within the same line.

File: client/src/app/interceptors/auth.interceptor.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        );
+      }
+      return throwError(() => new Error('Access Token expired: Login again'));
+     }));
+  }
+
+  private refreshAccessToken(): Observable<any> {
+
+    let user: User = User.fromJson(localStorage.getItem('user'));
```
Comment (paranerd) [inline client/src/app/interceptors/auth.interceptor.ts:64]: See above.

File: client/src/app/interceptors/index.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import {HTTP_INTERCEPTORS } from '@angular/common/http';
```
Comment (paranerd) [inline client/src/app/interceptors/index.ts:1]: Add spaces for consistency (`{ HTTP_INTERCEPTORS }`)

File: client/src/app/interceptors/index.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import {HTTP_INTERCEPTORS } from '@angular/common/http'; 
+import { AuthInterceptor } from './auth.interceptor';
+
+
+export const httpInterceptorProviders = [
+    {provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true}
```
Comment (paranerd) [inline client/src/app/interceptors/index.ts:6]: Maybe use multiple lines here for readability?

File: client/src/app/models/user.model.ts [hunk 10]
```diff
@@ -10,5 +11,11 @@ export class User {
     Object.assign(this, input);
 
     return this;
-}
+  }
+
+  static fromJson(input: any): User {
```
Comment (paranerd) [inline client/src/app/models/user.model.ts:16]: See above. This is already implemented.

**Review by paranerd — CHANGES_REQUESTED** [review]

File: client/angular.json (hunk 2 above)
Comment (maxkimambo) [reply client/angular.json:76]: CORS failures that you have been experiencing caused by : localhost:4200 > localhost:8080

Comment (paranerd) [reply client/angular.json:76]: I actually implemented a proxy to try to fix that but it didn't help so I removed it again.

File: client/src/app/interceptors/auth.interceptor.ts (hunk 4 above)
Comment (maxkimambo) [reply client/src/app/interceptors/auth.interceptor.ts:35]: I would like to have deserialise as a static function that returns user object. 
I didnt want to change that function hence i introduces fromJSon as in intermediate step to reduce blast radius of changes

Comment (paranerd) [reply client/src/app/interceptors/auth.interceptor.ts]: That would leave us with two functions basically doing the same thing. I think we should agree on one approach and stick to that.

I actually do like the static approach and parsing JSON within the static function. Reduces redundant code when using it.

**Review by paranerd — APPROVED** [review]

**Pull request merged**